File: _dsv/fromjson.py
import sys
import json
import argparse
from ._base import _Base

def flatten(d, sep='.', parent_key=None):
    data = {}
    if isinstance(d, dict):
        items = d.items()
    elif isinstance(d, list):
        items = enumerate(d)
    else:
        data[parent_key] = d
        return data

    for i, v in items:
        key = f"{parent_key}{sep}{i}" if parent_key else str(i)
        data.update(flatten(v, sep=sep, parent_key=key))
    return data

class fromjson(_Base):
    ''' convert from json '''
    parser = argparse.ArgumentParser()
    parser.add_argument('-f', '--flatten', nargs='?', const='.', help='flatten objects and arrays. (default seperator: %(const)s)')
    parser.add_argument('-s', '--slurp', action='store_true', help='determine header after reading all input')

    def parse_json(self, buffer, json_decoder=json.JSONDecoder()):
        try:
            utf8_buf = buffer.decode('utf8')
        except UnicodeDecodeError as e:
            utf8_buf = buffer[:e.start].decode('utf8')
            if not utf8_buf:
                raise
            remainder = buffer[e.start:]
        else:
            remainder = b''
        utf8_buf = utf8_buf.lstrip()

        value, index = json_decoder.raw_decode(utf8_buf)
        return value, utf8_buf[index:].lstrip().encode('utf8') + remainder

    def iter_json(self, file, chunk=8192):
        rest = b''
        while buf := file.read1(chunk):
            rest += buf
            try:
                while rest:
                    try:
                        value, rest = self.parse_json(rest)
                    except json.JSONDecodeError:
                        break
                    if isinstance(value, dict):
                        if self.opts.flatten:
                            value = flatten(value, self.opts.flatten)
                        yield value
                    else:
                        print('not a json object:', value, file=sys.stderr)
            except UnicodeDecodeError:
                break

        if rest:
            print('invalid json:', rest, file=sys.stderr)

    def calc_header(self, rows: list[dict]):
        self.header = list({key.encode('utf8'): None for row in rows for key in row})

    def process_file(self, file):
        self.determine_delimiters(b'')
        rows = self.iter_json(file)

        if self.opts.slurp:
            rows = list(rows)
            self.calc_header(rows)
            if super().on_header(self.header):
                return True

        for row in rows:
            if self.on_row(row):
                break

        self.on_eof()
        return ()

    def on_header(self, header):
        return self.on_row(header)

    def on_row(self, row):
        if self.header is None:
            self.calc_header([row])
            if super().on_header(self.header):
                return True

        row = [row.get(k.decode('utf8'), '') for k in self.header]
        row = [(x if isinstance(x, str) else json.dumps(x)).encode('utf8') for x in row]
        return super().on_row(row)

File: src/subcommands/fromjson.rs
use crate::utils::Break;
use anyhow::{Result, Context};
use crate::base::{self, Processor, Callbacks};
use bstr::{BStr, BString};
use std::process::ExitCode;
use std::io::Read;
use clap::Parser;
use serde_json;
use indexmap::{IndexMap, IndexSet};

#[derive(Parser, Debug)]
#[command(about = "convert from json")]
pub struct Opts {
    #[arg(short = 'f', long, num_args = 0..=1, help = "flatten objects and arrays")]
    flatten: Option<Option<String>>,
    #[arg(short = 's', long, help = "determine header after reading all input")]
    slurp: bool,
}

pub struct Handler {
    flatten: Option<String>,
    slurp: bool,
}

impl Handler {
    pub fn new(opts: Opts, _base: &mut base::Base) -> Result<Self> {
        Ok(Self {
            flatten: opts.flatten.map(|f| f.unwrap_or(".".into())),
            slurp: opts.slurp,
        })
    }
}

fn vec_iter<T>(vec: Vec<T>) -> impl Iterator<Item=(String, T)> {
    vec.into_iter().enumerate().map(|(i, v)| (format!("{i}"), v))
}

fn flatten<T: AsRef<[u8]>, I: Iterator<Item=(T, serde_json::Value)>>(
    iter: I,
    sep: &BStr,
    parent_key: Option<&BStr>,
    result: &mut IndexMap<BString, serde_json::Value>,
) {

    for (k, v) in iter {
        let key: BString = if let Some(p) = parent_key {
            bstr::concat([p, sep, k.as_ref()]).into()
        } else {
            k.as_ref().into()
        };
        match v {
            serde_json::Value::Object(map) => flatten(map.into_iter(), sep, Some(key.as_ref()), result),
            serde_json::Value::Array(vec) => flatten(vec_iter(vec), sep, Some(key.as_ref()), result),
            v => { result.insert(key, v); },
        }
    }
}

fn flatten_to_hashmap(map: serde_json::Map<String, serde_json::Value>, sep: &BStr) -> IndexMap<BString, serde_json::Value> {
    let mut result = IndexMap::new();
    flatten(map.into_iter(), sep, None, &mut result);
    result
}

impl Handler {

    fn process_json_row(
        &mut self,
        base: &mut base::Base,
        header: &IndexSet<&BStr>,
        row: &IndexMap<BString, serde_json::Value>
    ) -> Result<()> {
        let values = header.iter().map(|k| {
            row.get(*k)
                .map_or_else(String::new, |v| {
                    v.as_str().map_or_else(|| v.to_string(), |s| s.to_owned())
                })
                .into()
        }).collect();
        self.on_row(base, values)
    }

    fn calc_header<'a, I: IntoIterator<Item=&'a IndexMap<BString, T>>, T: 'a>(rows: I) -> IndexSet<&'a BStr> {
        rows.into_iter().flat_map(|row| row.keys()).map(|col| col.as_ref()).collect()
    }

    fn process_json<R: Read>(&mut self, file: R, base: &mut base::Base, do_callbacks: Callbacks) -> Result<()> {
        let sep = self.flatten.take();
        let mut stream = serde_json::Deserializer::from_reader(file)
            .into_iter()
            .map(|row| {
                row.map(|row| {
                    if let Some(sep) = sep.as_ref() {
                        flatten_to_hashmap(row, sep.as_bytes().into())
                    } else {
                        row.into_iter().map(|(k, v)| (k.into(), v)).collect()
                    }
                }).context("invalid json")
            });

        let do_header = do_callbacks.contains(Callbacks::ON_HEADER);
        let do_row = do_callbacks.contains(Callbacks::ON_ROW);
        let first_row = stream.next().ok_or(Break)??;

        if self.slurp {
            let mut rows = vec![first_row];
            for row in stream {
                rows.push(row?);
            }
            let header = Self::calc_header(rows.iter());
            if do_header {
                self.on_header(base, header.iter().map(|x| (*x).to_owned()).collect())?;
            }
            if do_row {
                for row in &rows {
                    self.process_json_row(base, &header, row)?;
                }
            }

        } else {
            // get the header only from the first row
            let header = Self::calc_header([&first_row]);
            if do_header {
                self.on_header(base, header.iter().map(|x| (*x).to_owned()).collect())?;
            }
            if do_row {
                self.process_json_row(base, &header, &first_row)?;
            }
            for row in stream {
                let row = row?;
                if do_row {
                    self.process_json_row(base, &header, &row)?;
                }
            }
        }

        Ok(())
    }
}

impl base::Processor for Handler {
    fn process_file<R: Read>(mut self, file: R, base: &mut base::Base, do_callbacks: Callbacks) -> anyhow::Result<ExitCode> {
        let ofs = self.determine_delimiters(b"".into(), &base.opts).1;
        base.on_ofs(ofs)?;
        // silence the break
        Break::is_break(self.process_json(file, base, do_callbacks))?;
        if do_callbacks.contains(Callbacks::ON_EOF) {
            return self.on_eof_detailed(base)
        }
        Ok(ExitCode::SUCCESS)
    }

}
